pkg/protocols/mqttprot/request.go

```diff
@@ -0,0 +1,164 @@
+	Client interface {
+		ClientID() string
+		UserName() string
+		Load(key interface{}) (value interface{}, ok bool)
+		Store(key interface{}, value interface{})
+		Delete(key interface{})
+	}
+
+	// PacketType contains supported MQTT packet type
+	PacketType int
+)
+
+const (
+	// ConnectType is MQTT packet type of connect
+	ConnectType PacketType = 1
+
+	// PublishType is MQTT packet type of publish
+	PublishType PacketType = 2
+
+	// DisconnectType is MQTT packet type of disconnect
+	DisconnectType PacketType = 3
+
+	// SubscribeType is MQTT packet type of subscribe
+	SubscribeType PacketType = 4
+
+	// UnsubscribeType is MQTT packet type of unsubscribe
+	UnsubscribeType PacketType = 5
+
+	// OtherType is all other MQTT packet type
+	OtherType PacketType = 99
+)
+
+var _ protocols.Request = (*Request)(nil)
+
+// NewRequest create new MQTT Request
+func NewRequest(packet packets.ControlPacket, client Client) *Request {
+	req := &Request{
+		client: client,
+		packet: packet,
+	}
+	switch p := packet.(type) {
+	case *packets.ConnectPacket:
+		req.packetType = ConnectType
+	case *packets.PublishPacket:
+		req.packetType = PublishType
+		req.payload = p.Payload
+	case *packets.DisconnectPacket:
+		req.packetType = DisconnectType
+	case *packets.SubscribePacket:
+		req.packetType = SubscribeType
+	case *packets.UnsubscribePacket:
+		req.packetType = UnsubscribeType
+	default:
+		req.packetType = OtherType
+	}
+	return req
+}
+
+// Client return MQTT request client
+func (r *Request) Client() Client {
+	return r.client
+}
+
+// PacketType return MQTT request packet type
+func (r *Request) PacketType() PacketType {
+	return r.packetType
+}
+
+// ConnectPacket return MQTT connect packet if PacketType is ConnectType
+func (r *Request) ConnectPacket() *packets.ConnectPacket {
+	return r.packet.(*packets.ConnectPacket)
+}
+
+// PublishPacket return MQTT publish packet if PacketType is PublishType
+func (r *Request) PublishPacket() *packets.PublishPacket {
+	return r.packet.(*packets.PublishPacket)
+}
+
+// DisconnectPacket return MQTT disconnect packet if PacketType is DisconnectType
+func (r *Request) DisconnectPacket() *packets.DisconnectPacket {
+	return r.packet.(*packets.DisconnectPacket)
+}
+
+// SubscribePacket return MQTT subscribe packet if PacketType is SubscribeType
+func (r *Request) SubscribePacket() *packets.SubscribePacket {
+	return r.packet.(*packets.SubscribePacket)
+}
+
+// UnsubscribePacket return MQTT unsubscribe packet if PacketType is UnsubscribeType
+func (r *Request) UnsubscribePacket() *packets.UnsubscribePacket {
+	return r.packet.(*packets.UnsubscribePacket)
+}
+
+// Header return MQTT request header
+func (r *Request) Header() protocols.Header {
+	// TODO: what header to return?
+	return nil
+}
+
+// SetPayload set the payload of the request to payload.
+func (r *Request) SetPayload(payload []byte) {
+	r.payload = payload
+	if r.packetType == PublishType {
+		r.packet.(*packets.PublishPacket).Payload = payload
+	}
```

Comment:
Is it make sense to set payload if packetType is not PublishType?
If not, we can panic here.